File: game/cave/cave_game.py
from audio_helper import AudioHelper
from cave.cave_resources import CaveResources
from cave.waiting_before_talking import WaitingBeforeTalking
from game_data import GameData
import global_state
from null_state import NullState
from phone_events import PhoneEvents
import pygame

class CaveGame:

    def __init__(self, context: GameData):
        self.ended = False
        self.rolling_score = context.forest_score
        self.context = context

        self.sounding_score = False
        self.cave_score_counter_start = None
        self._state = WaitingBeforeTalking(context)
        self._state.on_enter()

    def process_events(self, phone_events: PhoneEvents):
        next_state = self._state.process_events(phone_events)

        if next_state is not None:
            self._state.on_exit()
            if next_state == NullState:
                self.end()
            else:
                self._state = next_state(self.context)
                self._state.on_enter()

        if self.context.forest_score != self.rolling_score and not self.sounding_score:
            self.context.cave_score_counter_id = AudioHelper.play(CaveResources.score_counter, self.context.audio_port)
            self.cave_score_counter_start = global_state.frame_time
            self.sounding_score = True
        elif self.sounding_score and global_state.frame_time - self.cave_score_counter_start >= 0.1:
            self.context.cave_score_counter_id = AudioHelper.play(CaveResources.score_counter, self.context.audio_port)
            self.cave_score_counter_start = global_state.frame_time

    def render(self, screen):
        self._state.render(screen)

        if self.rolling_score < self.context.forest_score:
            self.rolling_score += min(10, self.context.forest_score - self.rolling_score)
        elif self.sounding_score:
            AudioHelper.stop(self.context.cave_score_counter_id, self.context.audio_port)
            self.sounding_score = False

        self.render_score(screen)

    def render_score(self, screen):
        x_score_right = 210 + 16 * 6  # Right edge position (6 digits max)
        y_score = 203
        x_space = 16

        score = int(self.rolling_score)
        if score == 0:
            screen.blit(CaveResources.score_font[0], (x_score_right - x_space, y_score))
            return

        digits = []
        temp = score
        while temp > 0:
            digits.append(temp % 10)
            temp //= 10

        for i, digit in enumerate(digits):
            screen.blit(CaveResources.score_font[digit], (x_score_right - x_space * (i + 1), y_score))

    def end(self):
        self._state.on_exit()
        if self.sounding_score:
            AudioHelper.stop(self.context.cave_score_counter_id, self.context.audio_port)
            self.sounding_score = False
        self.ended = True

File: game-web/src/cave/caveGame.ts
/** Port of game/cave/cave_game.py */

import { AudioHelper } from '../audioHelper';
import { CaveResources } from './caveResources';
import { WaitingBeforeTalking } from './waitingBeforeTalking';
import { NullState } from '../nullState';
import { globalState } from '../globalState';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';
import type { State } from '../state';

export class CaveGame {
  ended = false;

  private rollingScore: number;
  private context: GameData;
  private soundingScore = false;
  private caveScoreCounterStart = 0;
  private state: State;

  constructor(context: GameData) {
    this.context = context;
    this.rollingScore = context.forestScore;
    this.state = new WaitingBeforeTalking(context);
    this.state.onEnter();
  }

  processEvents(phoneEvents: PhoneEvents): void {
    const nextStateClass = this.state.processEvents(phoneEvents);

    if (nextStateClass !== null) {
      this.state.onExit();
      if (nextStateClass === NullState) {
        this.end();
      } else {
        this.state = new nextStateClass(this.context);
        this.state.onEnter();
      }
    }

    if (this.context.forestScore !== this.rollingScore && !this.soundingScore) {
      this.context.caveScoreCounterId = AudioHelper.play(CaveResources.scoreCounter);
      this.caveScoreCounterStart = globalState.frameTime;
      this.soundingScore = true;
    } else if (this.soundingScore && globalState.frameTime - this.caveScoreCounterStart >= 0.1) {
      this.context.caveScoreCounterId = AudioHelper.play(CaveResources.scoreCounter);
      this.caveScoreCounterStart = globalState.frameTime;
    }
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    this.state.render(ctx);

    if (this.rollingScore < this.context.forestScore) {
      this.rollingScore += Math.min(10, this.context.forestScore - this.rollingScore);
    } else if (this.soundingScore) {
      AudioHelper.stop(this.context.caveScoreCounterId);
      this.soundingScore = false;
    }

    this.renderScore(ctx);
  }

  private renderScore(ctx: OffscreenCanvasRenderingContext2D): void {
    const xScoreRight = 210 + 16 * 6;  // right edge position (6 digits max)
    const yScore = 203;
    const xSpace = 16;

    const score = Math.floor(this.rollingScore);
    if (score === 0) {
      ctx.drawImage(CaveResources.scoreFont[0], xScoreRight - xSpace, yScore);
      return;
    }

    const digits: number[] = [];
    let temp = score;
    while (temp > 0) {
      digits.push(temp % 10);
      temp = Math.floor(temp / 10);
    }

    for (let i = 0; i < digits.length; i++) {
      ctx.drawImage(CaveResources.scoreFont[digits[i]], xScoreRight - xSpace * (i + 1), yScore);
    }
  }

  private end(): void {
    this.state.onExit();
    if (this.soundingScore) {
      AudioHelper.stop(this.context.caveScoreCounterId);
      this.soundingScore = false;
    }
    this.ended = true;
  }
}
